[SPARK-37558][DOC] Improve spark sql cli document (#34821)
apache/spark: docs/sql-distributed-sql-engine-spark-sql-cli.md

    CLI options:
     -d,--define <key=value>          Variable substitution to apply to Hive
                                      commands. e.g. -d A=B or --define A=B
        --database <databasename>     Specify the database to use
     -e <quoted-query-string>         SQL from command line
     -f <filename>                    SQL from files
     -H,--help                        Print help information
        --hiveconf <property=value>   Use value for given property
        --hivevar <key=value>         Variable substitution to apply to Hive
                                      commands. e.g. --hivevar A=B
     -i <filename>                    Initialization SQL file
     -S,--silent                      Silent mode in interactive shell
     -v,--verbose                     Verbose mode (echo executed SQL to the
                                      console)

## The hiverc File

The Spark SQL CLI when invoked without the `-i` option will attempt to load `$HIVE_HOME/bin/.hiverc` and `$HOME/.hiverc` as initialization files.

## Spark SQL CLI Interactive Shell Commands

When `./bin/spark-sql` is run without either the `-e` or `-f` option, it enters interactive shell mode.

> Use `;` (semicolon) to terminate commands, but user can escape `;` by `\\;`. Comments in scripts can be specified using the `--` prefix.

Reviewer: What if `;` is not the end of the line? does it still terminate commands?

Reviewer: I think we should explain more about `;`
- what if it's inside a comment? `SELECT /* xxx; */ 1;`
- is `;` the only way to terminate commands? what if I type `SELECT 1` and then press enter?
- can we put multiple commands in a single line? `SELECT 1; SELECT 2;`

Author: How about current?

                                      commands. e.g. --hivevar A=B
     -i <filename>                    Initialization SQL file
     -S,--silent                      Silent mode in interactive shell
     -v,--verbose                     Verbose mode (echo executed SQL to the
                                      console)

## The hiverc File

The Spark SQL CLI when invoked without the `-i` option will attempt to load `$HIVE_HOME/bin/.hiverc` and `$HOME/.hiverc` as initialization files.

## Spark SQL CLI Interactive Shell Commands

When `$SPARK__HOME/bin/spark-sql` is run without either the `-e` or `-f` option, it enters interactive shell mode.
Use `;` (semicolon) to terminate commands, but user can escape `;` by `\\;`. Comments in scripts can be specified using the `--` prefix.

<table class="table">
<tr><th>Command</th><th>Description</th></tr>
<tr>
  <td><code>quit</code> <code>exit</code></td>
  <td>Use <code>quit</code> or <code>exit</code> to leave the interactive shell.</td>
</tr>
<tr>
  <td><code>!&lt;command&gt;</code></td>
  <td>Executes a shell command from the Spark SQL CLI shell.</td>
</tr>
<tr>
  <td><code>dfs &lt;dfs command&gt;</code></td>

>   <td>Executes a dfs command from the Hive shell.</td>

Reviewer 2: We should say what dfs means

Author: Add a dfs command link here

</tr>
<tr>
  <td><code>&lt;query string&gt;</code></td>
  <td>Executes a Spark SQL query and prints results to standard output.</td>
</tr>
<tr>
  <td><code>source &lt;filepath&gt;</code></td>
  <td>Executes a script file inside the CLI.</td>
</tr>
</table>

> ## Supported comment type

Reviewer 2: types

Author: Done

<table class="table">
<tr><th>Comment</th><th>Example</th></tr>
<tr>
  <td>simple comment</td>
  <td>
  <code>
      -- This is a simple comment.
      <br>
      SELECT 1;
  </code>
  </td>
</tr>
<tr>
  <td>bracketed comment</td>
  <td>
    <code>
        /* This is a bracketed comment. */
        <br>
        SELECT 1;
    </code>
  </td>
</tr>
<tr>
  <td>nested bracketed comment</td>
  <td>
    <code>
        /*  This is a /* nested bracketed comment*/ .*/
        <br>
        SELECT 1;
    </code>
  </td>
</tr>
</table>

## Examples

> See Variable Substitution for examples of using the hiveconf option.

Reviewer 2: Is this meant to be a link?

Author: Yea, remove this could be ok.

---
layout: global
title: Spark SQL CLI
displayTitle: Spark SQL CLI
license: |
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
 
     http://www.apache.org/licenses/LICENSE-2.0
 
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
---

* Table of contents
{:toc}

> The Spark SQL CLI is a convenient tool to run the Hive metastore service in local mode and execute

Reviewer 2: Is the note about Hive and Thrift needed here?
also "execute SQL queries"

Author: Thrift server doc here
https://spark.apache.org/docs/latest/sql-distributed-sql-engine.html#running-the-thrift-jdbcodbc-server

queries input from the command line. Note that the Spark SQL CLI cannot talk to the Thrift JDBC server.

To start the Spark SQL CLI, run the following in the Spark directory:

    ./bin/spark-sql

Configuration of Hive is done by placing your `hive-site.xml`, `core-site.xml` and `hdfs-site.xml` files in `conf/`.

## Spark SQL Command Line Options

You may run `./bin/spark-sql --help` for a complete list of all available options.

    CLI options:
     -d,--define <key=value>          Variable substitution to apply to Hive
                                      commands. e.g. -d A=B or --define A=B
        --database <databasename>     Specify the database to use
     -e <quoted-query-string>         SQL from command line
     -f <filename>                    SQL from files
     -H,--help                        Print help information
        --hiveconf <property=value>   Use value for given property

>         --hivevar <key=value>         Variable substitution to apply to Hive

Reviewer: what's the different between --hiveconf, --hivevar and --define?

Author: > what's the difference between --hiveconf, --hivevar and --define?

The hiveconf namespace and --hiveconf should be used to set Hive configuration values.
The hivevar namespace and --hivevar should be used to define user variables.
Setting user variables under the hiveconf namespace probably won't break anything, but isn't recommended.

The --hivevar variables are accessed using ${var-name}, while the --hiveconf are accessed ${hiveconf:var-name} inside hive.

                                      commands. e.g. --hivevar A=B
     -i <filename>                    Initialization SQL file
     -S,--silent                      Silent mode in interactive shell
     -v,--verbose                     Verbose mode (echo executed SQL to the
                                      console)

## The hiverc File

> The Spark SQL CLI when invoked without the `-i` option will attempt to load `$HIVE_HOME/bin/.hiverc` and `$HOME/.hiverc` as initialization files.

Reviewer 2: "When invoked without ..., the Spark SQL CLI will attempt ..."

Author: DOne

## Supported comment types

<table class="table">
<tr><th>Comment</th><th>Example</th></tr>
<tr>
  <td>simple comment</td>
  <td>
  <code>
      -- This is a simple comment.
      <br>
      SELECT 1;
  </code>
  </td>
</tr>
<tr>
  <td>bracketed comment</td>
  <td>
    <code>
        /* This is a bracketed comment. */
        <br>
        SELECT 1;
    </code>
  </td>
</tr>
<tr>
  <td>nested bracketed comment</td>
  <td>
    <code>
        /*  This is a /* nested bracketed comment*/ .*/
        <br>
        SELECT 1;
    </code>
  </td>
</tr>
</table>

## Spark SQL CLI Interactive Shell Commands

When `./bin/spark-sql` is run without either the `-e` or `-f` option, it enters interactive shell mode.
Use `;` (semicolon) to terminate commands. Notice:

> 1. CLI use `;` to terminate commands only when it's at the end of line and it's not escaped by `\\;`.

Reviewer 2: "The CLI uses `;` to ..."

Author: Done

> 2. `;` is the only way to terminate commands, if user type `SELECT 1` and press enter, console will just wait for input.

Reviewer 2: `;` is the only way to terminate commands. If the user types `SELECT 1` and presses enter, the console will just wait for input.

Author: DO ne

> 3. If user type multiple commands in one line like `SELECT 1; SELECT 2;`, commands `SELECT 1` and `SELECT 2` will be executed separatly.

Reviewer 2: If the user types multiple commands in one line like `SELECT 1; SELECT 2;`, then commands `SELECT 1` and `SELECT 2` will be executed separately.

> 4. If `;` in a simple comment 

Reviewer 2: If `;` appears in a simple comment, as in:

Author: Done

> 4. If `;` appears in a simple comment, as in: 

Reviewer: We don't need to mention `-- This is a comment;`, The `;` at the end means to terminate, which means we submit the query `-- This is comment` and does nothing. The same to `/* This is a comment contains ;`, we submit `/* This is a comment contains` and fails the parser.

The thing we need to mention additionally is what `;` means in the middle. e.g. `SELECT a = ';' FROM t;`, `-- comment;comment;`

Author: Like current?

   ```sql
   -- This is a comment;
   SELECT 1;
   ```

>    This comment line will just be ignored. If `;` in a bracketed command and not at the end of line,

Reviewer 2: then this comment line will be ignored. If `;` appears in a bracketed command

(and similarly for the other lines below)

Author: Done.

    ```sql
    /* This is a comment contains ';'. */
    SELECT 1;
    ```
  it will not terminate commands. If `;` in a bracketed command and in the end of line, 
     ```sql
     /* This is a comment contains ;
     */ SELECT 1;
     ```

>   it will terminate commands into  `/* This is a comment contains ` and `*/ SELECT 1`.

Reviewer: and we hit failures when executing `/* This is a comment contains `

Author: ```
spark-sql> /* This is a comment contains ;
Error in query:
Unclosed bracketed comment(line 1, pos 0)

== SQL ==
/* This is a comment contains ;
^^^

spark-sql>      */ SELECT 1;
Error in query:
extraneous input '*/' expecting {'(', 'ADD', 'ALTER', 'ANALYZE', 'CACHE', 'CLEAR', 'COMMENT', 'COMMIT', 'CREATE', 'DELETE', 'DESC', 'DESCRIBE', 'DFS', 'DROP', 'EXPLAIN', 'EXPORT', 'FROM', 'GRANT', 'IMPORT', 'INSERT', 'LIST', 'LOAD', 'LOCK', 'MAP', 'MERGE', 'MSCK', 'REDUCE', 'REFRESH', 'REPLACE', 'RESET', 'REVOKE', 'ROLLBACK', 'SELECT', 'SET', 'SHOW', 'START', 'TABLE', 'TRUNCATE', 'UNCACHE', 'UNLOCK', 'UPDATE', 'USE', 'VALUES', 'WITH'}(line 1, pos 5)

== SQL ==
     */ SELECT 1
-----^^^

spark-sql>
```

<table class="table">
<tr><th>Command</th><th>Description</th></tr>
<tr>

>   <td><code>quit</code> <code>exit</code></td>

Reviewer 2: put 'or' in between

Author: DOne

>   <td>Use <code>quit</code> or <code>exit</code> to leave the interactive shell.</td>

Reviewer 2: Just say "Exits the interactive shell" for consistency

Author: Done

</tr>
<tr>
  <td><code>!&lt;command&gt;</code></td>
  <td>Executes a shell command from the Spark SQL CLI shell.</td>
</tr>
<tr>
  <td><code>dfs &lt;dfs command&gt;</code></td>

>   <td>Executes a <a href="https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HDFSCommands.html#dfs">dfs command</a> from the Hive shell.</td>

Reviewer 2: Is this only for executing commands vs HDFS or would it work on other distributed storage?
We should say HDFS if it's HDFS.
Also, not "Hive shell" here right?

Author: Done, how about current?

</tr>
<tr>
  <td><code>&lt;query string&gt;</code></td>
  <td>Executes a Spark SQL query and prints results to standard output.</td>
</tr>
<tr>
  <td><code>source &lt;filepath&gt;</code></td>
  <td>Executes a script file inside the CLI.</td>
</tr>
</table>

## Examples

> Example of running a query from the command line

Reviewer 2: Put ':' at the end of lines like this here

Author: Done

## Spark SQL Command Line Options

You may run `./bin/spark-sql --help` for a complete list of all available options.

    CLI options:
     -d,--define <key=value>          Variable substitution to apply to Hive
                                      commands. e.g. -d A=B or --define A=B
        --database <databasename>     Specify the database to use
     -e <quoted-query-string>         SQL from command line
     -f <filename>                    SQL from files
     -H,--help                        Print help information
        --hiveconf <property=value>   Use value for given property
        --hivevar <key=value>         Variable substitution to apply to Hive
                                      commands. e.g. --hivevar A=B
     -i <filename>                    Initialization SQL file
     -S,--silent                      Silent mode in interactive shell
     -v,--verbose                     Verbose mode (echo executed SQL to the
                                      console)

## The hiverc File

When invoked without the `-i`, the Spark SQL CLI will attempt to load `$HIVE_HOME/bin/.hiverc` and `$HOME/.hiverc` as initialization files.

## Path interpretation

> Spark SQL CLI supports running SQL from initialization script file(`-i`) or normal SQL file(`-f`), If path url don't have a schema component, the path will be handled as local file.

Reviewer: ```suggestion
Spark SQL CLI supports running SQL from initialization script file(`-i`) or normal SQL file(`-f`), If path url don't have a scheme component, the path will be handled as local file.
```

> For example: `/path/to/spark-sql-cli.sql` equals to `file:///path/to/spark-sql-cli.sql`.

Reviewer: let's also mention that people can use Hadoop supported filesystem such as `s3://...`

<table class="table">
<tr><th>Command</th><th>Description</th></tr>
<tr>
  <td><code>quit</code> or <code>exit</code></td>
  <td>Exits the interactive shell.</td>
</tr>
<tr>
  <td><code>!&lt;command&gt;</code></td>
  <td>Executes a shell command from the Spark SQL CLI shell.</td>
</tr>
<tr>
  <td><code>dfs &lt;HDFS dfs command&gt;</code></td>
  <td>Executes a HDFS <a href="https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HDFSCommands.html#dfs">dfs command</a> from the Spark SQL CLI shell.</td>
</tr>
<tr>
  <td><code>&lt;query string&gt;</code></td>
  <td>Executes a Spark SQL query and prints results to standard output.</td>
</tr>
<tr>
  <td><code>source &lt;filepath&gt;</code></td>
  <td>Executes a script file inside the CLI.</td>
</tr>
</table>

## Examples

Example of running a query from the command line:

    ./bin/spark-sql -e 'SELECT COL FROM TBL'

Example of setting Hive configuration variables:

    ./bin/spark-sql -e 'SELECT COL FROM TBL' --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
    

> Example of setting Hive configuration variables:

Reviewer: ```suggestion
Example of setting Hive configuration variables and using it in the SQL query:
```

    ./bin/spark-sql -e 'SELECT ${hiveconf:aaa}' --hiveconf aaa=bbb --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
    spark-sql> SELECT ${aaa};
    bbb
    
Example of setting Hive variables substitution:
    
    ./bin/spark-sql --hivevar aaa=bbb --define ccc=ddd
    spark-sql> SELECT ${aaa}, ${ccc};
    bbb ddd

Example of dumping data out from a query into a file using silent mode:

    ./bin/spark-sql -S -e 'SELECT COL FROM TBL' > result.txt

Example of running a script non-interactively from local disk:

>     ./bin/spark-sql -f /path/to/spark-sql-script.sql

Reviewer: does it always mean local file? can `hadoop-site.xml` change the default file scheme?

Author: > does it always mean local file? can `hadoop-site.xml` change the default file scheme?

From the code, it can read hdfs file

```
  public int processFile(String fileName) throws IOException {
    Path path = new Path(fileName);
    Object fs;
    if (!path.toUri().isAbsolute()) {
      fs = FileSystem.getLocal(this.conf);
      path = ((FileSystem)fs).makeQualified(path);
    } else {
      fs = FileSystem.get(path.toUri(), this.conf);
    }

    BufferedReader bufferReader = null;
    boolean var5 = false;

    int rc;
    try {
      bufferReader = new BufferedReader(new InputStreamReader(((FileSystem)fs).open(path)));
      rc = this.processReader(bufferReader);
    } finally {
      IOUtils.closeStream(bufferReader);
    }

    return rc;
  }
```

Reviewer: then maybe let's remove this example, and add one more example below
```
    ./bin/spark-sql -f file:/path/to/spark-sql-script.sql
    ./bin/spark-sql -f hdfs://<namenode>:<port>/spark-sql-script.sql
    ./bin/spark-sql -f s3://mys3bucket/spark-sql-script.sql
```

Reviewer: please verify it locally first. I don't quite remember how local file scheme works

Author: > please verify it locally first. I don't quite remember how local file scheme works

For local
```
C02D45VVMD6T:spark yi.zhu$ ./bin/spark-sql -f /Users/yi.zhu/Documents/test.sql
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
21/12/13 14:52:22 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
21/12/13 14:52:24 WARN HiveConf: HiveConf of name hive.stats.jdbc.timeout does not exist
21/12/13 14:52:24 WARN HiveConf: HiveConf of name hive.stats.retries.wait does not exist
21/12/13 14:52:26 WARN ObjectStore: Version information not found in metastore. hive.metastore.schema.verification is not enabled so recording the schema version 2.3.0
21/12/13 14:52:26 WARN ObjectStore: setMetaStoreSchemaVersion called but recording version is disabled: version = 2.3.0, comment = Set by MetaStore yi.zhu@10.12.189.45
Spark master: local[*], Application Id: local-1639378343821
1
Time taken: 3.869 seconds, Fetched 1 row(s)
2
Time taken: 0.052 seconds, Fetched 1 row(s)
3
Time taken: 0.042 seconds, Fetched 1 row(s)
C02D45VVMD6T:spark yi.zhu$
```

File in hdfs's /user/yi.zhu/test.sql
```
(base) [yi.zhu@ip-xx-xx-xx-xx ~]$ /usr/share/spark/bin/spark-sql --master local -f /user/yi.zhu/test.sql
21/12/13 14:56:10 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
21/12/13 14:56:10 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
21/12/13 14:56:11 ERROR SparkSQLCLIDriver: Could not open input file for reading. (File file:/user/yi.zhu/test.sql does not exist)
```

```
(base) [yi.zhu@ip-10-130-21-35 ~]$ /usr/share/spark/bin/spark-sql --master local -f hdfs://R2/user/yi.zhu/test.sql
21/12/13 14:58:20 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
21/12/13 14:58:20 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
1
Time taken: 2.072 seconds, Fetched 1 row(s)
Error in query:
mismatched input 'seelct' expecting {'(', 'SELECT', 'FROM', 'ADD', 'DESC', 'WITH', 'VALUES', 'CREATE', 'TABLE', 'INSERT', 'DELETE', 'DESCRIBE', 'EXPLAIN', 'SHOW', 'USE', 'DROP', 'ALTER', 'MAP', 'SET', 'RESET', 'START', 'COMMIT', 'ROLLBACK', 'REDUCE', 'REFRESH', 'CLEAR', 'CACHE', 'UNCACHE', 'DFS', 'TRUNCATE', 'ANALYZE', 'LIST', 'REVOKE', 'GRANT', 'LOCK', 'UNLOCK', 'MSCK', 'EXPORT', 'IMPORT', 'LOAD'}(line 2, pos 0)

== SQL ==

seelct 2
^^^

(base) [yi.zhu@ip-xx-xx-xx-xx ~]$ cat test.sql
select 1;
seelct 2;
select 3;
```

  <td>Executes a script file inside the CLI.</td>
</tr>
</table>

## Examples

Example of running a query from the command line:

    ./bin/spark-sql -e 'SELECT COL FROM TBL'

Example of setting Hive configuration variables:

    ./bin/spark-sql -e 'SELECT COL FROM TBL' --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
    
Example of setting Hive configuration variables and using it in the SQL query:

    ./bin/spark-sql -e 'SELECT ${hiveconf:aaa}' --hiveconf aaa=bbb --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
    spark-sql> SELECT ${aaa};
    bbb
    
Example of setting Hive variables substitution:
    
    ./bin/spark-sql --hivevar aaa=bbb --define ccc=ddd
    spark-sql> SELECT ${aaa}, ${ccc};
    bbb ddd

Example of dumping data out from a query into a file using silent mode:

    ./bin/spark-sql -S -e 'SELECT COL FROM TBL' > result.txt

Example of running a script non-interactively from local disk:

    ./bin/spark-sql -f /path/to/spark-sql-script.sql
    ./bin/spark-sql -f file:///path/to/spark-sql-script.sql

Example of running a script non-interactively from a Hadoop supported filesystem:

    ./bin/spark-sql -f hdfs://<namenode>:<port>/spark-sql-script.sql

>     ./bin/spark-sql -f s3://mys3bucket/spark-sql-script.sql 

Reviewer: does this apply to the init file as well?

Author: > does this apply to the init file as well?

Yes， they call same method `processFile()`

Reviewer: then shall we have a individual section about path interpretation?

Author: > then shall we have a individual section about path interpretation?

Yea, necessary, it's not same as normal usage.

Author: How about current.

## Spark SQL Command Line Options

You may run `./bin/spark-sql --help` for a complete list of all available options.

    CLI options:
     -d,--define <key=value>          Variable substitution to apply to Hive
                                      commands. e.g. -d A=B or --define A=B
        --database <databasename>     Specify the database to use
     -e <quoted-query-string>         SQL from command line
     -f <filename>                    SQL from files
     -H,--help                        Print help information
        --hiveconf <property=value>   Use value for given property
        --hivevar <key=value>         Variable substitution to apply to Hive
                                      commands. e.g. --hivevar A=B
     -i <filename>                    Initialization SQL file
     -S,--silent                      Silent mode in interactive shell
     -v,--verbose                     Verbose mode (echo executed SQL to the
                                      console)

## The hiverc File

When invoked without the `-i`, the Spark SQL CLI will attempt to load `$HIVE_HOME/bin/.hiverc` and `$HOME/.hiverc` as initialization files.

## Path interpretation

> Spark SQL CLI support run SQL from initialization script file(`-i`) or normal SQL file(`-f`), If path is not absolute, the path will be handled as local file.

Reviewer: ```suggestion
Spark SQL CLI supports running SQL from initialization script file(`-i`) or normal SQL file(`-f`), If path is not absolute, the path will be handled as local file.
```

Reviewer: The word `absolute` is a bit vague. Is `/path/to/data` absolute? Is `path/to/date` absolute?

Author: `If path don't have a schema component ` seems can be more clear?

</tr>
<tr>
  <td><code>&lt;query string&gt;</code></td>
  <td>Executes a Spark SQL query and prints results to standard output.</td>
</tr>
<tr>
  <td><code>source &lt;filepath&gt;</code></td>
  <td>Executes a script file inside the CLI.</td>
</tr>
</table>

## Examples

Example of running a query from the command line:

    ./bin/spark-sql -e 'SELECT COL FROM TBL'

Example of setting Hive configuration variables:

    ./bin/spark-sql -e 'SELECT COL FROM TBL' --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
    
Example of setting Hive configuration variables and using it in the SQL query:

    ./bin/spark-sql -e 'SELECT ${hiveconf:aaa}' --hiveconf aaa=bbb --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
    spark-sql> SELECT ${aaa};
    bbb
    
Example of setting Hive variables substitution:
    
    ./bin/spark-sql --hivevar aaa=bbb --define ccc=ddd
    spark-sql> SELECT ${aaa}, ${ccc};
    bbb ddd

Example of dumping data out from a query into a file using silent mode:

    ./bin/spark-sql -S -e 'SELECT COL FROM TBL' > result.txt

> Example of running a script non-interactively from local disk:
> 
>     ./bin/spark-sql -f /path/to/spark-sql-script.sql

Reviewer: ```suggestion
Example of running a script non-interactively:

    ./bin/spark-sql -f /path/to/spark-sql-script.sql
```
This should be good enough, as we have already explained how path is interpreted.

---
layout: global
title: Spark SQL CLI
displayTitle: Spark SQL CLI
license: |
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
 
     http://www.apache.org/licenses/LICENSE-2.0
 
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
---

* Table of contents
{:toc}

The Spark SQL CLI is a convenient tool to run the Hive metastore service in local mode and execute SQL
queries input from the command line. Note that the Spark SQL CLI cannot talk to the Thrift JDBC server.

To start the Spark SQL CLI, run the following in the Spark directory:

    ./bin/spark-sql

> Configuration of Hive is done by placing your `hive-site.xml`, `core-site.xml` and `hdfs-site.xml` files in `conf/`.

Reviewer: does it mean we can even configure a remote HMS?

Author: > does it mean we can even configure a remote HMS?

In my experience, we can do like this to add some execution related hive conf and hadoop conf.

Reviewer: > The Spark SQL CLI is a convenient tool to run the Hive metastore service in local mode ...

Then this is not accurate. Maybe `run the Hive metastore service (in local mode by default) ...`

Reviewer: ping @AngersZhuuuu

Reviewer: This is not addressed yet.

## Supported comment types

<table class="table">
<tr><th>Comment</th><th>Example</th></tr>
<tr>
  <td>simple comment</td>
  <td>
  <code>
      -- This is a simple comment.
      <br>
      SELECT 1;
  </code>
  </td>
</tr>
<tr>
  <td>bracketed comment</td>
  <td>
    <code>
        /* This is a bracketed comment. */
        <br>
        SELECT 1;
    </code>
  </td>
</tr>
<tr>
  <td>nested bracketed comment</td>
  <td>
    <code>
        /*  This is a /* nested bracketed comment*/ .*/
        <br>
        SELECT 1;
    </code>
  </td>
</tr>
</table>

## Spark SQL CLI Interactive Shell Commands

When `./bin/spark-sql` is run without either the `-e` or `-f` option, it enters interactive shell mode.
Use `;` (semicolon) to terminate commands. Notice:
1. The CLI use `;` to terminate commands only when it's at the end of line, and it's not escaped by `\\;`.
2. `;` is the only way to terminate commands. If the user types `SELECT 1` and presses enter, the console will just wait for input.
3. If the user types multiple commands in one line like `SELECT 1; SELECT 2;`, the commands `SELECT 1` and `SELECT 2` will be executed separatly.

> 4. If `;` appears in a simple comment, as in: 

Reviewer: ```suggestion
4. If `;` appears in a simple comment, but not at the end:
```
